### PR TITLE
[1.1] fix: Broken cache on Windows

### PR DESCRIPTION
**Pull request by serverwentdown**
(Backport to 1.1) (Original PR: #4531)

The previous implementation would fail to install packages on Windows
because it creates a `Path` starting with a slash. Such a `Path` is
invalid on Windows. Instead, use the utility function url_to_path.

# Pull Request Check List

Resolves: #4479 #4535

- [x] Added **tests** for changed code
- [x] Updated **documentation** for changed code. (N/A)

### CODE REVIEW

**Comment by GoodiesHQ** [comment]
I can at least confirm that this solves my issue. All I did was install poetry using `install-poetry.py` and manually change `%APPDATA%\pypoetry\venv\Lib\site-packages\poetry\installation\executor.py` to match the pull request. I am now able to successfully install packages on windows. It's bizarre, I would imagine this would be a rampant issue with more people on Windows...

**Comment by barskern** [comment]
Also having this issue. Hopefully it gets merged soon, seems like an uncontroversial change.

**Comment by serverwentdown** [comment]
Ping @sdispater

**Comment by ditw11mhs** [comment]
Can't wait for this merge, my current workaround is just to delete the cache folder and running poetry install again, well it work but doin it and waiting for the file to be deleted is a pain in the noggin :'

**Comment by Greatness7** [comment]
Can confirm that the current release of poetry was totally broken on my windows system, and this PR fixed it. Seems like something that should be merged ASAP!

**Comment by f0ff886f** [comment]
>Thanks a lot for the PR! I will check with the teams we have there as this brings a lot of pain to everyone!

@MadJlzz Is there any way people outside of poetry core can help?

Poetry has been broken without manual intervention since June 9 (150+ days) for Windows users, is there anything we can do to get a fix rolled out?

**Comment by Luttik** [comment]
@f0ff886f I've tried to get a fork from poetry with @serverwentdown deployed but that turned out to be more than a 30-minute job 😔. I think that we need to at least get the `get-poetry.py` script to work with the fork for it to be usable as an alternative... I just don't have the time at this point to figure out everything that needs to change to get both the release workflow to work aswell as the `get-poetry.py`. It seems like we at least need an alternative to https://pypi.org/pypi/poetry/json but I've gotten no clue how that one is generated.

I did reserve poetry-windows-fix on pypi with the @serverwentdown 's fix online already. If anybody wants to continue the work I can transfer it or add you as a contibruter to both the github and pypi repo.

**Comment by serverwentdown** [comment]
@Luttik if you'd like to install my branch of poetry, you can run the following `pip` command (note that this install method is not recommended by the documentation due to the potential for dependency conflicts, but if you don't have any other tools installed globally, this should be fine):

```
git+https://github.com/serverwentdown/poetry.git@1.1-fix-cache-on-windows
```

**Comment by neersighted** [comment]
> 
> 
> @Luttik if you'd like to install my branch of poetry, you can run the following `pip` command (note that this install method is not recommended by the documentation due to the potential for dependency conflicts, but if you don't have any other tools installed globally, this should be fine):
> 
> ```
> git+https://github.com/serverwentdown/poetry.git@1.1-fix-cache-on-windows
> ```

You can also use `install-poetry.py` with the `--git` option to do it the recommended way e.g. `curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/install-poetry.py | python - --git git+https://github.com/serverwentdown/poetry.git@1.1-fix-cache-on-windows`

**Comment by Luttik** [comment]
@serverwentdown I realise that this is possible (I assume that it is even better to do this with pipx). I am however very aware of my lack of awareness regarding the downsides and why the get-poetry script is what it is...

**Review by neersighted — CHANGES_REQUESTED** [review]
Looks good to merge once the PR is rebased, the format complies with our lints, and the checks pass.

**Review by neersighted — APPROVED** [review]

**Comment by MadJlzz** [comment]
> > Thanks a lot for the PR! I will check with the teams we have there as this brings a lot of pain to everyone!
> 
> @MadJlzz Is there any way people outside of poetry core can help?
> 
> Poetry has been broken without manual intervention since June 9 (150+ days) for Windows users, is there anything we can do to get a fix rolled out?

Sorry for the misunderstanding ; I am not from the Poetry core team. I meant "check with the team at my company" to help them out. 😢 

Thanks @neersighted for merging this out and @serverwentdown for the fix!

**Comment by Luttik** [comment]
FYI @neersighted the command should be without `git+` it seems like the script prepends the url with `git+` already.

so it bocomes: `curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/install-poetry.py | python - --git https://github.com/serverwentdown/poetry.git@1.1-fix-cache-on-windows`

Could you edit your comment since I believe many use it as a reference.

**Comment by neersighted** [comment]
> 
> 
> FYI @neersighted the command should be without `git+` it seems like the script prepends the url with `git+` already.
> 
> so it bocomes: `curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/install-poetry.py | python - --git https://github.com/serverwentdown/poetry.git@1.1-fix-cache-on-windows`
> 
> Could you edit your comment since I believe many use it as a reference.

My bad. Thanks!

**Comment by damarvin** [comment]
just to mention, I tried this curl+install as given, but still see the leading backslash → ValueError  File \C:\Users\...

**Comment by serverwentdown** [comment]
@damarvin That's the command for Linux. For Windows: 

```
(Invoke-WebRequest -Uri https://raw.githubusercontent.com/python-poetry/poetry/master/install-poetry.py -UseBasicParsing).Content | python - --git git+https://github.com/python-poetry/poetry.git@master
```

**Comment by damarvin** [comment]
Thanks for the hint! (but...)
I really prefer the cmd.exe over powershell.
What is wrong with curl? The given
`curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/install-poetry.py | python - --git https://github.com/serverwentdown/poetry.git@1.1-fix-cache-on-windows` works for me, _Your_ line (started in powershell) left me with
`ERROR: Invalid requirement: 'git+git+https://github.com/python-poetry/poetry.git@1.1'`

The point is, poetry.git@1.1-fix-cache-on-windows seemingly does not fix the "leading backslash" issue.
I'm trying this in a `venv`, so sure it is newly installed (and for this issue I accustomed to clean the poetry caches).

**Comment by serverwentdown** [comment]
I believe the `curl` command wouldn't work on Windows. That's why you're not getting the updated version.

I have updated my comment to fix the PowerShell command.

**Comment by damarvin** [comment]
_SIDE NOTE:
curl/wget is fine (like via cygwin), powershell is not the daily terminal for many developers under Windows—I stay with cmd.exe.
May I propose, you also provide a command line w/o PowerShell at https://python-poetry.org/docs/master/#installation?
Btw, the instruction shell lines are crappy, at least with Firefox they end with "... python -", no chance to scroll further, no copy button.
And `%APPDATA%\Roaming\Python\Scripts` should read `%APPDATA%\Roaming\Python\Scripts`.
And you should mention the parallel `%APPDATA%\Roaming\pypoetry`.
And you should mention that poetry.exe under former is only a symbolic link to latter._

`poetry install` from poetry.git@1.1 tells now

    [SolverProblemError]
    Because <project> depends on pyproject-flake8 (*) which doesn't match any versions, version solving failed.

While going back to

    pip install poetry
    
I see now

    Attempting uninstall: poetry
      Found existing installation: poetry 1.1.11
      Uninstalling poetry-1.1.11:
        Successfully uninstalled poetry-1.1.11
    ERROR: Could not install packages due to an OSError: [WinError 5] Access is denied: 'C:\\Users\\<user>\\AppData\\Local\\Temp\\pip-uninstall-km8mnbvv\\poetry.exe'

I mean you should rely on introduced ways, not fiddling in the user's system.
I really much prefer calling in a venv `pip install poetry` over the over-complicated new `install-poetry.py`.
I like the ideas of poetry over setuptools, but please keep it simple and do not waste the users' time.

**Comment by damarvin** [comment]
https://github.com/python-poetry/poetry/pull/4836 fixes it.
Thanks a lot, @finswimmer !

**Comment by Red-Eyed** [comment]
Just wonder, couldn't `Path().as_posix()` be used?

**Comment by serverwentdown** [comment]
@Red-Eyed It wouldn't directly address the mistake made by the previous author.

**Pull request merged**
